File: rules-interpreter/src/test/java/CompareTest.java
import io.github.iamrenny.ruleflow.Workflow;
import io.github.iamrenny.ruleflow.vo.WorkflowResult;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.Map;

class CompareTest {

    @Test
    public void givenStringWhenOperatingLtOperationMustGoOk() {
        String workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'x_is_lesser' x < '7.53' return block
                default allow
            end
        """;

        Workflow ruleEngine = new Workflow(workflow);
        WorkflowResult expectedResult = new WorkflowResult("test", "dummy", "x_is_lesser", "block");
        WorkflowResult result = ruleEngine.evaluate(Map.of("x", "7.52.2.19911"));

        Assertions.assertEquals(expectedResult, result);
    }

    @Test
    public void givenStringWhenOperatingLteOperationMustGoOk() {
        String workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'x_is_lesser' x <= '7.53' return block
                default allow
            end
        """;

        Workflow ruleEngine = new Workflow(workflow);
        WorkflowResult expectedResult = new WorkflowResult("test", "dummy", "x_is_lesser", "block");
        WorkflowResult result = ruleEngine.evaluate(Map.of("x", "7.52.2.19911"));

        Assertions.assertEquals(expectedResult, result);
    }

    @Test
    public void givenStringWhenOperatingGtOperationMustGoOk() {
        String workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'x_greater' x > '7.53' return block
                default allow
            end
        """;

        Workflow ruleEngine = new Workflow(workflow);
        WorkflowResult expectedResult = new WorkflowResult("test", "dummy", "x_greater", "block");
        WorkflowResult result = ruleEngine.evaluate(Map.of("x", "7.53.2.19911"));

        Assertions.assertEquals(expectedResult, result);
    }

    @Test
    public void givenStringWhenOperatingGteOperationMustGoOk() {
        String workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'x_greater' x >= '7.53' return block
                default allow
            end
        """;

        Workflow ruleEngine = new Workflow(workflow);
        WorkflowResult expectedResult = new WorkflowResult("test", "dummy", "x_greater", "block");
        WorkflowResult result = ruleEngine.evaluate(Map.of("x", "7.53.2.19911"));

        Assertions.assertEquals(expectedResult, result);
    }

    @Test
    public void givenStringWhenOperatingNeOperationMustGoOk() {
        String workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'x_greater' x <> '7.53' return block
                default allow
            end
        """;

        Workflow ruleEngine = new Workflow(workflow);
        WorkflowResult expectedResult = new WorkflowResult("test", "dummy", "x_greater", "block");
        WorkflowResult result = ruleEngine.evaluate(Map.of("x", "7.52"));

        Assertions.assertEquals(expectedResult, result);
    }

    @Test
    public void givenDateTimeWithZoneWhenComparedMustMatch() {
        String workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'dt_match' x = '2024-06-01T12:30Z' return block
                default allow
            end
        """;

        Workflow ruleEngine = new Workflow(workflow);
        WorkflowResult expectedResult = new WorkflowResult("test", "dummy", "dt_match", "block");
        WorkflowResult result = ruleEngine.evaluate(Map.of("x", "2024-06-01T12:30Z"));

        Assertions.assertEquals(expectedResult, result);
    }
}

File: rules-interpreter/src/test/kotlin/CompareTest.kt
import com.rappi.fraud.rules.parser.WorkflowEvaluator
import com.rappi.fraud.rules.parser.vo.WorkflowEvaluatorResult
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class CompareTest {

    @Test
    fun `given a string when operating lt operation must go ok`() {
        val workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'x_is_lesser' x < '7.53'  return block
                default allow
            end
        """

        val ruleEngine = WorkflowEvaluator(workflow)
        Assertions.assertEquals(
            WorkflowEvaluatorResult( "test", "dummy", "x_is_lesser", "block"),
            ruleEngine.evaluate(
                mapOf(
                    "x" to "7.52.2.19911"
                )
            )
        )
    }

    @Test
    fun `given a string when operating lte operation must go ok`() {
        val workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'x_is_lesser' x <= '7.53'  return block
                default allow
            end
        """

        val ruleEngine = WorkflowEvaluator(workflow)
        Assertions.assertEquals(
            WorkflowEvaluatorResult( "test", "dummy", "x_is_lesser", "block"),
            ruleEngine.evaluate(
                mapOf(
                    "x" to "7.52.2.19911"
                )
            )
        )
    }
    @Test
    fun `given a string when operating gt operation must go ok`() {
        val workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'x_greater' x > '7.53'  return block
                default allow
            end
        """

        val ruleEngine = WorkflowEvaluator(workflow)
        Assertions.assertEquals(
            WorkflowEvaluatorResult( "test", "dummy", "x_greater", "block"),
            ruleEngine.evaluate(
                mapOf(
                    "x" to "7.53.2.19911"
                )
            )
        )
    }

    @Test
    fun `given a string when operating gte operation must go ok`() {
        val workflow = """
            workflow 'test'
                ruleset 'dummy'
                    'x_greater' x >= '7.53'  return block
                default allow
            end
        """

        val ruleEngine = WorkflowEvaluator(workflow)
        Assertions.assertEquals(
            WorkflowEvaluatorResult( "test", "dummy", "x_greater", "block"),
            ruleEngine.evaluate(
                mapOf(
                    "x" to "7.53.2.19911"
                )
            )
        )
    }
}